### PR TITLE
added function to give a player_id and get a dict with their season stats

### PR DESCRIPTION
**Pull request by mattalexhoward**

### CODE REVIEW

File: src/util/players_util.py [hunk 1]
```diff
@@ -227,3 +227,20 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
     return (ftsy_prj, value)
+
+"""
+Given a player_id this will return a dictionary of this players stats for the season.
+Example: get_player_stats_dict(2544) will return a dict of dict of stat value pairs.
+{2017: "player_id" : 2544, "pts" : 9000 ... , 2016 : "player_id" : 2544 , "pts" : 5 }
+"""
+def get_player_season_stats_dict(player_id):
+    player_season_stats_dict = defaultdict(list)
+    query = {"player_id" : player_id}
```
Comment (jcgoodru) [inline src/util/players_util.py]: Can you use f.player_id instead of the raw string?

File: src/util/players_util.py [hunk 2]
```diff
@@ -227,3 +227,20 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
     value = ftsy_prj/ftsy_pts_last_10 if ftsy_pts_last_10 > 0 else 1
     return (ftsy_prj, value)
+
+"""
+Given a player_id this will return a dictionary of this players stats for the season.
+Example: get_player_stats_dict(2544) will return a dict of dict of stat value pairs.
+{2017: "player_id" : 2544, "pts" : 9000 ... , 2016 : "player_id" : 2544 , "pts" : 5 }
+"""
+def get_player_season_stats_dict(player_id):
+    player_season_stats_dict = defaultdict(list)
```
Comment (jcgoodru) [inline src/util/players_util.py]: what does this do?  If it's a dict of dicts why is it defaultdict-ing to a list?

You can use:
```
player_season_stats_dict = {}
```

File: src/util/players_util.py [hunk 3]
```diff
@@ -227,3 +227,20 @@ def calculate_fantasy_points(player_id, opp_team_id):
     ftsy_prj = ftsy_prj * recent_form
     value = ftsy_prj/ftsy_pts_last_10 if ftsy_pts_last_10 > 0 else 1
     return (ftsy_prj, value)
+
+"""
+Given a player_id this will return a dictionary of this players stats for the season.
+Example: get_player_stats_dict(2544) will return a dict of dict of stat value pairs.
+{2017: "player_id" : 2544, "pts" : 9000 ... , 2016 : "player_id" : 2544 , "pts" : 5 }
```
Comment (jcgoodru) [inline src/util/players_util.py]: If this is here, it should reflect the type that's being returned

File: src/util/players_util.py [hunk 4]
```diff
@@ -227,3 +227,20 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+"""
+def get_player_season_stats_dict(player_id):
+    player_season_stats_dict = defaultdict(list)
+    query = {"player_id" : player_id}
+    player_record_cursor = connection.PlayerSeasonStatsRecord.find(query)
+    
+    for each in player_record_cursor:
```
Comment (jcgoodru) [inline src/util/players_util.py]: Can you use "rec" or "record" instead of "each?"  Each is not descriptive, and looks like a reserved word

**Comment by RyanDanielOMara** [comment]
For the stats included in the dict, could you put the same stats in the same order as the yearly stats at this link? 

http://www.nba.com/players/lebron/james/2544

**Comment by mattalexhoward** [comment]
@RyanDanielOMara @jcgoodru Updated what you requested.

File: src/util/players_util.py [hunk 5]
```diff
@@ -236,3 +237,67 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+"""
+Given a player_id this will return a list containig player stats.
+Returns [header, career, season_dict]
+'header' is a list of the stat names in the order they appear in the career and season lists
+'career' is a list of stats for the players complete career (up until 2008)
+'season_dict' is a dict that contains season based player stats
+Ex: season_dict[2017] returns a list of stats for ONLY 2017
+"""
+def get_player_season_stats(player_id):
+    query = {"player_id" : player_id}
```
Comment (jcgoodru) [inline src/util/players_util.py]: can you use f.player_id

File: src/util/players_util.py [hunk 6]
```diff
@@ -236,3 +237,67 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+        season = dict(each.items())[f.season] #gets the current season out of the Cursor object
+        season_list = [None] * 21 #creates list to be big enough for all headers to fit on
+
+        for stat_tuple in each.items(): #for each stat, insert it into the list where we want it
+            ind = lookup_index_of_stat(stat_tuple[0])
+            if ind != -1: #some stats we dont care about. They have an 'index' of -1. We throw them away.
+                # add the stats to both the list for this season and the career stats
+                season_list[ind] = (stat_tuple[1]) 
+                career_stats[ind] += stat_tuple[1]
+        season_list[5] = season_list[3] / season_list[4] * 100
```
Comment (jcgoodru) [inline src/util/players_util.py]: Does this produce an int, since all 3 numbers are ints?  Maybe you need to add 100.0 * list[3] / list[4]?

Not sure tbh, but can you verify?

File: src/util/players_util.py [hunk 7]
```diff
@@ -236,3 +237,67 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+          ]
+    def lookup_index_of_stat(stat):
+        try:
+            return header.index(stat)
+        except ValueError:
+            return -1;
+
+    career_stats = [0]*21
+    season_dict = {}
+    for each in player_record_cursor:
```
Comment (jcgoodru) [inline src/util/players_util.py]: Can you use rec instead of each?  Each isn't descriptive of what you're iterating over, and it sounds like a reserved word

File: src/util/players_util.py [hunk 8]
```diff
@@ -236,3 +237,67 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+def get_player_season_stats(player_id):
+    query = {"player_id" : player_id}
+    player_record_cursor = connection.PlayerSeasonStatsRecord.find(query)
+    
+    """ 
+    Will take stat "pts" and return index of that stat in the list to be returned.
+    This is a helper function to order the stats list how it was requested by thefront end team.
+    """ 
+    header = [f.games_played,
+               f.minutes,
```
Comment (jcgoodru) [inline src/util/players_util.py]: We can probably take this out, it might not be relevant to know that LeBron played 3263 minutes in 2010.  We could average it out but even then it might not be a very useful stat

File: src/util/players_util.py [hunk 9]
```diff
@@ -236,3 +237,67 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+        try:
+            return header.index(stat)
+        except ValueError:
+            return -1;
+
+    career_stats = [0]*21
+    season_dict = {}
+    for each in player_record_cursor:
+
+        season = dict(each.items())[f.season] #gets the current season out of the Cursor object
```
Comment (jcgoodru) [inline src/util/players_util.py]: Not exactly sure what this does, and I don't see where it's being used

**Comment by jcgoodru** [comment]
I forgot to mention:

Can you make season_dict a list of rows instead?  For instance, ```seasons = [[2017 stuff], [2016 stuff], [2015 stuff], ... ]```
This lets us display the data as-is (instead of re-run processing it before displaying it to the page)

File: src/util/players_util.py (hunk 6 above)
Comment (mattalexhoward) [reply src/util/players_util.py]: this will produce a float

File: src/util/players_util.py (hunk 9 above)
Comment (mattalexhoward) [reply src/util/players_util.py]: Are yo asking about line 289? If so  
>gets the current season out of the Cursor object
each record is a season year. This retrieved the season year from inside of the record so we know what year it is so it can be added to the dict under that same season.

**Comment by mattalexhoward** [comment]
>This lets us display the data as-is (instead of re-run processing it before displaying it to the page)

Wat mean by dis?

File: src/util/players_util.py (hunk 9 above)
Comment (jcgoodru) [reply src/util/players_util.py]: Oh, I see.  I didn't see it being used, but you're using it in line 301.

You can simplify this into ```each.season```.  And it might be better to use it in line 301 directly (no need for a separate variable).  If you instantiate variables near the spot they're being used, it's easier to see how they're being used.

Also, if you use a term like ```rec``` or ```record``` (or something indicating what the item is) instead of ```each```, it can be easier for a reviewer to see optimizations such as this

File: src/util/players_util.py (hunk 6 above)
Comment (jcgoodru) [reply src/util/players_util.py]: Did you verify this?  Integer division only yields integers in Python2

File: src/util/players_util.py [hunk 10]
```diff
@@ -236,3 +237,67 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+        season_list = [None] * 21 #creates list to be big enough for all headers to fit on
+
+        for stat_tuple in each.items(): #for each stat, insert it into the list where we want it
```
Comment (jcgoodru) [inline src/util/players_util.py]: Instead of iterating across the PlayerSeasonStatRecord, can you iterate across the header_list instead?

File: src/util/players_util.py (hunk 6 above)
Comment (mattalexhoward) [reply src/util/players_util.py]: >Did you verify this? Integer division only yields integers in Python2
 Write

no i just told you this for shits and giggles

Comment (jcgoodru) [reply src/util/players_util.py]: This is wrong and needs to be fixed.
I'm not sure how you verified this.
Additionally, I tested it and it doesn't work.
I tested this programatically, despite the fact that it's so obviously broken that it doesn't need testing to verify.


```env/bin/python src/broken_code.py```

contents of broken_code.py (copy-pasted the includes and your broken function):
```
import string
import math
import operator
import unicodedata
import pymongo

from datetime import date, datetime
from database.connection import DATABASE_NAME, connection
from database.tables.fields import Fields as f
from nba_py import team as nba_team
from nba_py import player as nba_player
from database.tables.league.player_prediction import PlayerPredictionRecord
from database.tables.league.player_season_stats import PlayerSeasonStatsRecord


"""
Given a player_id this will return a list containig player stats.
Returns [header, career, season_dict]
'header' is a list of the stat names in the order they appear in the career and season lists
'career' is a list of stats for the players complete career (up until 2008)
'season_dict' is a dict that contains season based player stats
Ex: season_dict[2017] returns a list of stats for ONLY 2017
"""
def get_player_season_stats(player_id):
    query = {"player_id" : player_id}
    player_record_cursor = connection.PlayerSeasonStatsRecord.find(query)
    
    """ 
    Will take stat "pts" and return index of that stat in the list to be returned.
    This is a helper function to order the stats list how it was requested by thefront end team.
    """ 
    header = [f.games_played,
               f.minutes,
               f.pts,
               f.fgm,
               f.fga,
               "fg%",
               f.fg3m,
               f.fg3a,
               "3p%",
               f.ftm, 
               f.fta,
               "ft%",
               f.oreb,
               f.dreb,
               f.reb,
               f.ast,
               f.tov,
               f.stl,
               f.blk,
               f.fouls, 
               f.plus_minus,
          ]
    def lookup_index_of_stat(stat):
        try:
            return header.index(stat)
        except ValueError:
            return -1;

    career_stats = [0]*21
    season_dict = {}
    for each in player_record_cursor:

        season = dict(each.items())[f.season] #gets the current season out of the Cursor object
        season_list = [None] * 21 #creates list to be big enough for all headers to fit on

        for stat_tuple in each.items(): #for each stat, insert it into the list where we want it
            ind = lookup_index_of_stat(stat_tuple[0])
            if ind != -1: #some stats we dont care about. They have an 'index' of -1. We throw them away.
                # add the stats to both the list for this season and the career stats
                season_list[ind] = (stat_tuple[1]) 
                career_stats[ind] += stat_tuple[1]
        season_list[5] = season_list[3] / season_list[4] * 100
        season_list[8] = season_list[6] / season_list[7] * 100
        season_list[11] = season_list[9] / season_list[10] * 100
        season_dict[season] = season_list

    return [header, career_stats, season_dict]



print get_player_season_stats(2544)
```

Results of the broken code:

```
[['games_played', 'minutes', 'pts', 'fgm', 'fga', 'fg%', 'fg3m', 'fg3a', '3p%', 'ftm', 'fta', 'ft%', 'oreb', 'dreb', 'reb', 'ast', 'tov', 'stl', 'blk', 'fouls', 'plus_minus'], [166, 6120, 4337, 1655, 3071, 0, 239, 694, 0, 788, 1113, 0, 228, 1100, 1328, 1297, 627, 216, 112, 308, 1083], {2016: [74, 2796, 1954, 736, 1344, 0, 124, 342, 0, 358, 531, 0, 97, 542, 639, 646, 303, 92, 44, 134, 483], 2017: [16, 616, 463, 182, 311, 0, 28, 70, 0, 71, 91, 0, 20, 104, 124, 137, 75, 20, 19, 31, -18], 2015: [76, 2708, 1920, 737, 1416, 0, 87, 282, 0, 359, 491, 0, 111, 454, 565, 514, 249, 104, 49, 143, 618]}]

```

Comment (jcgoodru) [reply src/util/players_util.py]: If you're having a hard time, I can write this function for you

**Comment by mattalexhoward** [comment]
@RyanDanielOMara @jcgoodru made the requested changes

File: src/util/players_util.py [hunk 11]
```diff
@@ -236,3 +237,105 @@ def calculate_fantasy_points(player_id, opp_team_id):
     ftsy_prj = round(ftsy_prj * recent_form, 1)
     value = ftsy_prj/ftsy_pts_last_10 if ftsy_pts_last_10 > 0 else 1
     return (ftsy_prj, value)
+
+"""
+Given a player_id this will return a list containig player stats.
+Returns [header, career, list_list_season]
+'header' is a list of the stat names in the order they appear in the career and season lists
+'career' is a list of stats for the players complete career (up until 2008)
```
Comment (jcgoodru) [inline src/util/players_util.py]: Can you remove ```(up until 2008)```?

The end year is determined by how far someone backfilled
(ex: for me, it's up until 2015, and for someone it may go further back than 2008)

File: src/util/players_util.py [hunk 12]
```diff
@@ -236,3 +237,105 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+               f.dreb,
+               f.reb,
+               f.ast,
+               f.tov,
+               f.stl,
+               f.blk,
+               f.fouls, 
+               f.plus_minus,
+          ]
+    def lookup_index_of_stat(stat):
```
Comment (jcgoodru) [inline src/util/players_util.py]: This isn't being used anymore, can you remove it?

File: src/util/players_util.py [hunk 13]
```diff
@@ -236,3 +237,105 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+Given a player_id this will return a list containig player stats.
+Returns [header, career, list_list_season]
+'header' is a list of the stat names in the order they appear in the career and season lists
+'career' is a list of stats for the players complete career (up until 2008)
+'list_list_season' is a list of season lists. The season list contains [2017, pts, fga,...]
+The 'list_list_season' contains contains [[2017, pts, fgm, ...], [2016, pts, fgm, ...], [2015, pts, fgm, ...], ... ]
+
+"""
+def get_player_season_stats(player_id):
+    query = {"player_id" : player_id}
```
Comment (jcgoodru) [inline src/util/players_util.py]: Can you change this to ```f.player_id``` instead of ```"player_id"```?

File: src/util/players_util.py [hunk 14]
```diff
@@ -236,3 +237,105 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+    player_record_cursor = connection.PlayerSeasonStatsRecord.find(query)
+    
+    """ 
+    Will take stat "pts" and return index of that stat in the list to be returned.
+    This is a helper function to order the stats list how it was requested by thefront end team.
+    """ 
+    header = [f.games_played,
```
Comment (jcgoodru) [inline src/util/players_util.py]: The header needs to contain ```f.season```, since this will show on the website.  Otherwise the columns of the header won't match the columns of the seasonal data
(ex: the first column is games played in the header, but it's the season-year in the list_list_season rows)

File: src/util/players_util.py [hunk 15]
```diff
@@ -236,3 +237,105 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+               f.plus_minus,
+          ]
+    def lookup_index_of_stat(stat):
+        try:
+            return header.index(stat)
+        except ValueError:
+            return -1;
+    career_stats = [0]*len(header)
+    career_stats[0] = None
+    list_list_season = []  #contains [[2017, pts, fgm, ...], [2016, pts, fgm, ...], [2015, pts, fgm, ...], ... ]
```
Comment (jcgoodru) [inline src/util/players_util.py]: Can you rename this ```all_season_stats``` or something a bit more descriptive of what's inside of it?  ```list_list_season``` is somewhat vague

File: src/util/players_util.py [hunk 16]
```diff
@@ -236,3 +237,105 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+            ## If this header is one we need to compute, compute it
+            if stat == 'fg%': 
+                value = round(100.0 * rec.fgm / rec.fga , 1) if rec.fga else '-'
+
+            elif stat == '3p%':
+                value = round(100.0 * rec.fg3m / rec.fg3a, 1) if rec.fga else '-'
+
+            elif stat == 'ft%':
+                value = round(100.0 * rec.ftm / rec.fta, 1) if rec.fga else '-'
+                value = 0
```
Comment (jcgoodru) [inline src/util/players_util.py]: ```value = 0```
Why is this here?

File: src/util/players_util.py [hunk 17]
```diff
@@ -236,3 +237,105 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+                value = round(100.0 * rec.fgm / rec.fga , 1) if rec.fga else '-'
+
+            elif stat == '3p%':
+                value = round(100.0 * rec.fg3m / rec.fg3a, 1) if rec.fga else '-'
+
+            elif stat == 'ft%':
+                value = round(100.0 * rec.ftm / rec.fta, 1) if rec.fga else '-'
+                value = 0 
+
+            elif stat == f.games_played:
```
Comment (jcgoodru) [inline src/util/players_util.py]: When you add f.season to the beginning of the header, you can replace this like and the one after it with this:

```
elif stat in [ f.games_played, f.season ]:
    value = getattr(rec, stat)
```

File: src/util/players_util.py [hunk 18]
```diff
@@ -236,3 +237,105 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+
+            elif stat == f.season: 
+                value = rec.season
+
+            else: 
+                value = getattr(rec, stat)
+                career_stats[len(season_list)] += getattr(rec, stat)
+            season_list.append(value)
+
+        list_list_season.insert(0, season_list)
```
Comment (jcgoodru) [inline src/util/players_util.py]: After sorting (mentioned in an above comment), you can just ```thingy.append(season_list)```

File: src/util/players_util.py (hunk 13 above)
Comment (mattalexhoward) [reply src/util/players_util.py]: you asked me to do this 3 times lol

File: src/util/players_util.py [hunk 19]
```diff
@@ -236,3 +237,105 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+    def lookup_index_of_stat(stat):
+        try:
+            return header.index(stat)
+        except ValueError:
+            return -1;
+    career_stats = [0]*len(header)
+    career_stats[0] = None
+    list_list_season = []  #contains [[2017, pts, fgm, ...], [2016, pts, fgm, ...], [2015, pts, fgm, ...], ... ]
+
+    for rec in player_record_cursor:
```
Comment (jcgoodru) [inline src/util/players_util.py]: You can sort this by season.  I don't think it's guaranteed to be sorted in any order unless you do this.

```
player_record_cursor = sorted(player_record_cursor, key=lambda rec : rec.season, reverse=True)
for rec in player_record_cursor:
    ## Do stuff
```

File: src/util/players_util.py [hunk 20]
```diff
@@ -236,3 +237,105 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+            if season[percent_stat] == '-':
+                average_percent = 0
+            else:
+                average_percent += season[percent_stat]
+            count += 1
+
+        career_stats[percent_stat] = round(average_percent / count, 1) if average_percent != 0 else '-'
+
+
+#     print career_stats
```
Comment (jcgoodru) [inline src/util/players_util.py]: Yank this out

File: src/util/players_util.py (hunk 17 above)
Comment (mattalexhoward) [reply src/util/players_util.py]: this makes adding to the career stats a little more complex.

File: src/util/players_util.py [hunk 21]
```diff
@@ -236,3 +237,105 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+               f.blk,
+               f.fouls, 
+               f.plus_minus,
+          ]
+    def lookup_index_of_stat(stat):
+        try:
+            return header.index(stat)
+        except ValueError:
+            return -1;
+    career_stats = [0]*len(header)
```
Comment (jcgoodru) [inline src/util/players_util.py]: What if you use a dict instead?  This can make your code shorter and easier to understand

```
## (1) Remove everything involving career_stats
## (2) Use this, instead of career_stats
career_stats_dict = { col : 0 for col in header }

## (3) In the elif stat == f.games_played (line 310 as of now)
        career_stats_dict[stat] += value ## This adds games played to the total

## (4) In the else clause (line 317 as of now)
        career_stats_dict[stat] += value ## This adds whatever integer stat to the total

## (5) Remove lines 324 - 336

## (6) Add this:

## Normalize all the stats
career_stats = [ round(1.0 * career_stats_dict[stat] / career_stats_dict[f.games_played], 1) for stat in headers ]

## Fix the stats that shouldn't be normalized
csd = career_stats_dict
career_stats[ft_pct_idx] = round(100.0 * csd[f.ftm] / csd[f.fta], 1) if csd[f.fta] else '-'
career_stats[fg_pct_idx] = round(100.0 * csd[f.fgm] / csd[f.fga], 1) if csd[f.fga] else '-'
career_stats[fg3_pct_idx] = round(100.0 * csd[f.fg3m] / csd[f.fg3a], 1) if csd[f.fg3a] else '-'
career_stats[games_played_index] = csd[f.games_played]
```

```

File: src/util/players_util.py [hunk 22]
```diff
@@ -236,3 +237,94 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+
+            else: 
+                value = round(1.0 * getattr(rec, stat) / games_played, 1)
+                career_stats[len(season_list)] += getattr(rec, stat)
+            season_list.append(value)
+
+        list_of_season_lists.append(season_list)
+
+
+    # average out these percentage stats for the career stats
```
Comment (jcgoodru) [inline src/util/players_util.py]: Can you remove this and implement the method I mentioned a couple commits ago?

This way of doing it is susceptible to bugs, whereas the other is much simpler and about 20 lines shorter

Comment (jcgoodru) [reply src/util/players_util.py]: You can remove lines 277, 278, 302, 309, 316-328 and insert this instead:

Comment (jcgoodru) [reply src/util/players_util.py]: Current line 277 becomes this:
```
career_stats_dict = { col : 0 for col in header }
```

278 gets removed

302, 309 become this (same indentation level as now):
```
career_stats_dict[stat] += value
```

316-328 get deleted

Insert this instead:
```
## Normalize all the stats
career_stats = [ round(1.0 * career_stats_dict[stat] / career_stats_dict[f.games_played], 1) for stat in headers ]

## Fix the stats that shouldn't be normalized
csd = career_stats_dict
career_stats[ft_pct_idx] = round(100.0 * csd[f.ftm] / csd[f.fta], 1) if csd[f.fta] else '-'
career_stats[fg_pct_idx] = round(100.0 * csd[f.fgm] / csd[f.fga], 1) if csd[f.fga] else '-'
career_stats[fg3_pct_idx] = round(100.0 * csd[f.fg3m] / csd[f.fg3a], 1) if csd[f.fg3a] else '-'
career_stats[games_played_index] = csd[f.games_played]
```

I put csd there because I got tired of writing career_stats_dict.  You can remove that if you wish, but it doesn't slow down the program to have it in there (since python assignment uses references)

File: src/util/players_util.py [hunk 23]
```diff
@@ -236,3 +237,94 @@ def calculate_fantasy_points(player_id, opp_team_id):
...
+    list_of_season_lists = []  #contains [[2017, pts, fgm, ...], [2016, pts, fgm, ...], [2015, pts, fgm, ...], ... ]
+
+    player_record_cursor = sorted(player_record_cursor, key=lambda rec : rec.season, reverse=True)
+    for rec in player_record_cursor:
+        ## we need a new list for each player season
+        season_list = []
+        for stat in header:
+
+            ## get games_played so we only have to retrieve it once for normalization
+            games_played = rec.games_played
```
Comment (jcgoodru) [inline src/util/players_util.py]: This needs to come out, it doesn't do anything

File: src/util/players_util.py (hunk 22 above)
Comment (mattalexhoward) [reply src/util/players_util.py]: remove what?
312?

**Comment by mattalexhoward** [comment]
#72
